File: oscilador armonico simple/oas_solucion_forzado.m
clear variables
close all


% Estos parametros describen la ecuacion del sistema:
%
%    m*x'' = -k*x - b*x' + F0*cos(w_ext*t)
%
par.m     = 1;      % masa, kg
par.k     = 1;      % constante elástica, N m^-1
par.b     = 1;      % coeficiente disipativo, N s m^-1 = kg m s^-2 s m^-1 = kg s^-1
par.w_ext = .8;      % frecuencia angular de la fuerza externa, Hz
par.F0    = 1;      % amplitud de la fuerza externa, N

par.tau_ext = 2*pi/par.w_ext;   % como referencia obtenemos el período de la f. ext

% Poniendo todos m, k, y w_ext iguales a 1 obtenemos muy fácilmente el régimen 
% de resonancia del sistema.

% Estos parametros permiten simplificar la ecuacion del sistema (simplemente
% ganamos un coeficiente 1 en el término de derivada segunda):
%
%   x'' + gamma*x' + w0^2*x = a0*cos(w_ext*t)
%
% Para esto, todos los parámetros se normalizan según la masa. Pasamos de 5
% parámetros a 4. (La frecuencia de la fuerza externa no se ve modificada, dado
% que está dentro de un coseno).
par.w0_sq = par.k  / par.m;     % frecuencia angular natural del sistema en
                                % ausencia de disipación, elevada al cuadrado
par.gamma = par.b  / par.m;
par.a0    = par.F0 / par.m;

% Escribimos la solución analítica en función del tiempo. Primero definimos los
% puntos del tiempo donde la evaluaremos:
t_min     = 0;               % Por claridad escribimos los tiempos inicial,
t_max     = 2*par.tau_ext;   % final (dado por dos períodos de f. ext),
N_samples = 100000;          % y la cantidad de samples temporales.
% 1000 samples entre 0 y 2*tau_ext segundos, equiespaciadas (frec. de sampleo = 1000 / 2*tau_ext))
t = linspace(t_min, t_max, N_samples);

% Luego definimos los coeficientes para la solución particular (omitimos la
% homogénea ya que solo existe durante un período de tiempo finito).

aux = (par.w0_sq - par.w_ext^2)^2 + par.gamma^2 * par.w_ext^2; % este factor es
                                % común a las amplitudes A y B, lo escribimos
                                % una sola vez para trabajar menos y tener menos
                                % chance de cometer errores.
A = par.gamma*par.w_ext*par.a0 / aux;
B = (par.w0_sq - par.w_ext^2) * par.a0 / aux;

% Luego escribimos la solución particular, su velocidad y aceleración. Acá vemos
% por primera vez una de las características principales de MatLab/Octave. t es
% un vector, y MatLab/Octave permite evaluar la solución para todos los puntos
% del tiempo en una sola línea de código. (Internamente realiza la evaluación
% para cada punto).
xp =            A*sin(par.w_ext*t) + B*cos(par.w_ext*t);        % x_p(t)
vp = par.w_ext*(A*cos(par.w_ext*t) - B*sin(par.w_ext*t));       % v_p(t)
ap = -xp*par.w_ext^2;                                           % a_p(t)

% Luego escribimos las fuerzas involucradas
F_res = -par.k*xp;                  % Fuerza elástica vs. t
F_dis = -par.b*vp;                  % Fuerza disipativa vs. t
F_ext = par.F0*cos(par.w_ext*t);    % Fuerza externa vs. t

% Graficamos todo lo que obtuvimos hasta ahora

figure
    subplot(4,1,1)
        plot(t, xp)
        ylabel('x_p(t) [m]')
    subplot(4,1,2)
        plot(t, vp)
        ylabel('v_p(t) [m/s]')
    subplot(4,1,3)
        plot(t, ap)
        ylabel('a_p(t) [m/s^2]')
    subplot(4,1,4)
        hold all
        plot(t, F_res)
        plot(t, F_dis)
        plot(t, F_ext)
        ylabel('Fuerzas [N]')
        legend('Elástica', 'Disipativa', 'Externa')
        xlabel('Tiempo (s)')
        box on

% Ahora analizamos la energia. Calculamos primero el trabajo "instantáneo" que
% realiza cada fuerza. Usamos dos nuevas funcionalidades de Matlab/Octave.
% Primero evaluamos la fuerza en todos los puntos del tiempo excepto el último
% (ahora veremos por qué) usando la notación "(1:end-1)". Esto se lee de la
% siguiente manera: la expresión a la izquierda de ":" indica el primer elemento
% a evaluar (1), y la expresión a la derecha indica el último (end-1, esto
% quiere decir el primero antes del último elemento del vector, es decir el
% segundo contando desde atrás; end-2 es el antepenúltimo, etc.). Luego usamos
% diff(xp) para obtener el desplazamiento realizado por el sistema en cada punto
% del tiempo. El operador ".*" permite multiplicar dos vectores de iguales
% dimensiones elemento a elemento, y guardar cada resultado en un nuevo vector
% de igual tamaño a los originales. De este modo obtenemos los diferenciales
% discretos asociados a cada instante del tiempo (su precisión obviamente
% dependerá de la frecuencia de muestreo establecida).

dW_res = F_res(1:end-1) .* diff(xp);
dW_dis = F_dis(1:end-1) .* diff(xp);
dW_ext = F_ext(1:end-1) .* diff(xp);

figure
    % Ponemos como referencia la fuerza externa en un panel superior
    subplot(5,1,1)
        plot(t, F_ext)
        set(gca, 'YTickLabel', '', 'XTickLabel', '')
        ylabel('x_p(t)')

    % Aquí mostramos los diferenciales de trabajo de cada fuerza
    subplot(5,1, 2:5)
        hold all
        plot(t(1:end-1), dW_res)
        plot(t(1:end-1), dW_dis)
        plot(t(1:end-1), dW_ext)
        plot(t, zeros(size(t)), 'k--')  % linea de referencia en cero, debe ir
                                        % al final para no interferir con la
                                        % leyenda
        xlabel('Tiempo (s)')
        ylabel('dW [N m]')
        legend('Elástica', 'Disipativa', 'Externa')
        box on

% Interesantemente, vemos que la fuerza elástica realiza trabajo positivo y
% negativo alternadamente en cada medio ciclo de la fuerza externa, de modo que 
% el trabajo neto asociado en un ciclo completo es nulo. También vemos que las
% fuerzas disipativa y externa realizan trabajo opuesto en cada instante del
% tiempo. Cada una por separado realiza trabajo neto distinto de cero en un
% ciclo completo (positivo para la externa, negativo para la disipativa, como es
% de esperar*), pero en cada momento se cancelan mutuamente. Esto es
% importantísimo ya que asegura que la amplitud del sistema no diverja incluso
% en la frecuencia de resonancia (donde la transferencia de energía asociada a
% la fuerza externa es máxima).

% * Analizar qué ocurre cambiando la frecuencia de la fuerza externa

% Finalmente podemos calcular el trabajo acumulado desde el tiempo inicial hasta
% cada instante del tiempo. Para esto usamos la función cumsum, que suma todos
% los elementos de un vector y los coloca en un nuevo vector y coloca las sumas
% parciales en cada elemento del vector resultado. Ejemplo, si x = [ 1, 3, 4 ]
% obtenemos cumsum(x) = [ 1, 4, 8 ].

W_res = [ 0, cumsum(dW_res) ];      % con esta notación agregamos un 0 al principio de cada vector
W_dis = [ 0, cumsum(dW_dis) ];
W_ext = [ 0, cumsum(dW_ext) ];

figure
    hold all
    plot(t, W_res)
    plot(t, W_dis)
    plot(t, W_ext)
    plot(t, W_dis+W_ext)
    plot(t, zeros(size(t)), 'k--')  % linea de referencia en cero

    xlabel('Tiempo (s)')
    ylabel('W [N m]')
    legend('Elástica', 'Disipativa', 'Externa', 'Dis. + Ext.', 'Location', 'northwest')

    % Ademas imprimimos los valores finales de cada tabajo
    text(t(end), W_res(end), num2str(W_res(end)))
    text(t(end), W_dis(end), num2str(W_dis(end)))
    text(t(end), W_ext(end), num2str(W_ext(end)))

fprintf('Trabajo neto al final de la simulación, en J (chequear el valor de t_max):\n')
fprintf('- Elástico:    %.2f\n', W_res(end))
fprintf('- Disipativo:  %.2f\n', W_dis(end))
fprintf('- Externo:     %.2f\n', W_ext(end))
fprintf('- Dis. + Ext.: %.2f\n', W_dis(end)+W_ext(end))

fprintf('Potencia media en un ciclo (¿cuánto debe valer t_max?):\n')
fprintf('- Elástica:    %.2f\n',  W_res(end) / t_max)
fprintf('- Disipativa:  %.2f\n',  W_dis(end) / t_max)
fprintf('- Externa:     %.2f\n',  W_ext(end) / t_max)
fprintf('- Dis. + Ext.: %.2f\n', (W_dis(end)+W_ext(end)) / t_max)

File: oscilador_armonico_simple/oas_solucion_forzado.py
# -*- coding: utf-8 -*-
import matplotlib.pyplot as plt
import numpy as np

"""
Amplitudes para la solucion particular de un oscilador armonico simple
amortiguado, forzado por una fuerza armonica simple

"""

""" 1. parametros """

# parametros del sistema


m = 2.          # masa
k = 1 / 8e-6    # constante elastica
b = 20          # coef disipacion

Fext = 1        # amplitud fuerza externa
wext = 25      # frecuencia fuerza externa

n_ciclos = 3
t_min    = 0
t_max    = 2 * np.pi / wext * n_ciclos      # de este modo, t_max corresponde a la cantidad
                                            # especificada de ciclos de oscilacion
n_puntos = n_ciclos*10000                   # cuantos samples por ciclo


""" 2. solucion particular """

#-- parametros derivados
a0    = Fext / m                # aceleracion del forzante
gamma = b / m                   # disipacion por unidad de masa
w0_sq = k / m                   # constante elastica por unidad de masa (es la
                                # frecuencia natural al cuadrado)

# uso variables auxiliares para escribir las amplitudes de la solucion
# particular, de este modo evito repetir formulas y cometer errores
dif_w_sq    = w0_sq - wext**2 
denominador = dif_w_sq**2 + (wext*gamma)**2

A = a0 * (dif_w_sq)   / denominador
B = a0 * wext * gamma / denominador 

#-- solucion
t  = np.linspace(t_min, t_max, n_puntos)                # tiempo
xp =   A*np.cos(wext*t) + B*np.sin(wext*t)              # posicion
vp = (-A*np.sin(wext*t) + B*np.cos(wext*t))*wext        # velocidad
ap = -xp*wext**2                                        # aceleracion

# fuerzas presentes
F_res = -k*xp                                           # Fuerza elástica vs. t
F_dis = -b*vp                                           # Fuerza disipativa vs. t
F_ext = Fext*np.cos(wext*t)                             # Fuerza externa vs. t

# trabajo de las fuerzas (diferenciales)
d_xp   = np.concatenate(( np.array([0]), np.diff(xp) )) # diferencial para la posicion, chequear que sea chico
dW_res = F_res * d_xp                                   # \
dW_dis = F_dis * d_xp                                   # |_ diferenciales para cada fuerza
dW_ext = F_ext * d_xp                                   # /  

# trabajo acumulado por cada fuerza
W_res = np.cumsum(dW_res)
W_dis = np.cumsum(dW_dis)
W_ext = np.cumsum(dW_ext)

""" --- 3. graficos """


fig1, ax1 = plt.subplots(3, 1)
fig1.set_size_inches(6, 9)

ax1[0].plot(t, F_ext, label='Fext')
ax1[0].set(ylabel='$F_{ext}$ [mks]', xticklabels = [])

ax1[1].plot(t, xp, label = 'x_p(t)')
ax1[1].set(ylabel='$x_p$ [mks]', xticklabels = [])

ax1[2].plot(t, vp, label = 'v_p(t)')
ax1[2].set(xlabel = '$t$ [mks]', ylabel = '$v_p$ [mks]')

fig1.show()


fig2, ax2 = plt.subplots(4, 1)
fig2.set_size_inches(6, 8)

ax2[0].plot(t, F_res)
ax2[0].plot(t, F_dis)
ax2[0].plot(t, F_ext)
ax2[0].set(ylabel='$F$ [mks]', xticklabels = [])

ax2[1].plot(t, dW_res, label='res')
ax2[1].plot(t, dW_dis, label='dis')
ax2[1].plot(t, dW_ext, label='ext')
ax2[1].set(ylabel='$dW$ [mks]', xticklabels = [])

ax2[2].plot(t, W_res)
ax2[2].plot(t, W_dis)
ax2[2].plot(t, W_ext)
ax2[2].set(ylabel='$W$ [mks]', xticklabels = [])

ax2[3].plot(t, W_ext + W_dis, color='firebrick', label='ext+dis')
ax2[3].set(ylabel='$W$ [mks]',xlabel='$t$ [mks]')

fig2.legend()
fig2.show()

